curvefs/src/mds/topology/topology_manager.cpp

```diff
@@ -693,55 +693,74 @@ void TopologyManager::CreatePartitions(const CreatePartitionRequest *request,
-            LOG(ERROR) << "Select metaServer failed when create partition"
-                       << "zoneId = " << zoneId << "erro code = " << ret;
+            LOG(ERROR) << "Select Pool failed when create partition"
+                       << ", erro code = " << ret;
             return ret;
         }
-        MetaServer metaServer;
-        if (topology_->GetMetaServer(metaServerId, &metaServer)) {
-            metaServerAddrs.emplace(
-                metaServer.GetInternalHostIp() + ":" +
-                std::to_string(metaServer.GetInternalPort()));
-        } else {
-            LOG(ERROR) << "Get metaserver failed.";
-            return TopoStatusCode::TOPO_METASERVER_NOT_FOUND;
+
+        Pool pool;
+        if (!topology_->GetPool(poolId, &pool)) {
+            LOG(ERROR) << "Get pool failed.";
+            return TopoStatusCode::TOPO_POOL_NOT_FOUND;
         }
-        metaServerIds.emplace(metaServerId);
-    }
 
-    if (metaServerAddrs.size() != replicaNum) {
-        LOG(ERROR) << "Have invalid metaserver number when create copyset."
-                   << " metaserver number = " << metaServerAddrs.size()
-                   << ", replicanum = " << replicaNum;
-    }
+        replicaNum = pool.GetReplicaNum();
+        int needZoneNum = replicaNum;
+        std::set<ZoneIdType> zones;
+        std::set<ZoneIdType> unavailableZones;
+        do {
+            ret = topology_->ChooseZonesInPool(poolId, &zones, unavailableZones,
+                                            needZoneNum);
+            if (TopoStatusCode::TOPO_OK != ret) {
+                LOG(ERROR) << "Select Zone failed when create partition"
+                        << "poolId = " << poolId << ", erro code = " << ret;
+                unavailablePools.emplace(poolId);
+                metaServerIds.clear();
+                metaServerIds.clear();
+                break;
+            }
+
+            for (auto zoneId : zones) {
+                MetaServerIdType metaServerId;
+                ret = topology_->ChooseSingleMetaServerInZone(zoneId,
+                                                        &metaServerId);
+                if (TopoStatusCode::TOPO_OK != ret) {
+                    LOG(WARNING) << "Select metaServer failed when create"
+                                 << " partition zoneId = " << zoneId
+                                 << ", erro code = " << ret;
+                    unavailableZones.emplace(zoneId);
+                    continue;
+                }
+
+                MetaServer metaServer;
+                if (topology_->GetMetaServer(metaServerId, &metaServer)) {
+                    metaServerAddrs.emplace(
+                        metaServer.GetInternalHostIp() + ":" +
+                        std::to_string(metaServer.GetInternalPort()));
+                } else {
+                    LOG(ERROR) << "Get metaserver failed.";
+                    return TopoStatusCode::TOPO_METASERVER_NOT_FOUND;
```

Comment:
mark this metaserver unavailable, and select another one?

Reply:
> mark this metaserver unavailable, and select another one?

This situation must some errors in cluster.